crates/ruff/src/jupyter/notebook.rs

```diff
@@ -0,0 +1,291 @@
+                                },
+                                Range::default(),
+                            )
+                        } else {
+                            Diagnostic::new(
+                                SyntaxError {
+                                    message: format!(
+                                        "Expected a Jupyter Notebook (.{JUPYTER_NOTEBOOK_EXT} extension), \
+                                    which must be internally stored as JSON, \
+                                    but found a Python source file: {err}"
+                                    ),
+                                },
+                                Range::default(),
+                            )
+                        }
+                    }
+                    Category::Data => {
+                        // We could try to read the schema version here but if this fails it's
+                        // a bug anyway
+                        Diagnostic::new(
+                            SyntaxError {
+                                message: format!(
+                                    "This file does not match the schema expected of Jupyter Notebooks: {err}"
+                                ),
+                            },
+                            Range::default(),
+                        )
+                    }
+                }
+            }));
+        }
+    };
+
+    // v4 is what everybody uses
+    if notebook.nbformat != 4 {
+        // bail because we should have already failed at the json schema stage
+        return Err(Box::new(Diagnostic::new(
+            SyntaxError {
+                message: format!(
+                    "Expected Jupyter Notebook format 4, found {}",
+                    notebook.nbformat
+                ),
+            },
+            Range::default(),
+        )));
+    }
+    if !notebook
+        .metadata
+        .language_info
+        .as_ref()
+        .map_or(true, |language| language.name == "python")
+    {
+        debug!(
+            "Skipping {} because it's not a python notebook",
+            path.display()
+        );
+        return Ok(None);
+    }
+
+    Ok(Some(notebook))
+}
+
+/// Concatenates all cells into a single virtual file and builds an index that maps the content
+/// to notebook cell locations
+pub fn concat_notebook(notebook: &JupyterNotebook) -> (String, JupyterIndex) {
+    let mut jupyter_index = JupyterIndex {
+        // Enter a line number (1-based), get back the cell (1-based)
+        // 0 index is just padding
+        row_to_cell: vec![0],
+        // Enter a line number (1-based), get back the row number in the cell (1-based)
+        // 0 index is just padding
+        row_to_row_in_cell: vec![0],
+    };
+    let size_hint = notebook
+        .cells
+        .iter()
+        .filter(|cell| cell.cell_type == CellType::Code)
+        .count();
+
+    let mut contents = Vec::with_capacity(size_hint);
```

Comment:
Did you profile if setting the `size_hint` outperforms iterating over the code cells?

Reply:
no i did that based on charlie's comment above; i currently also don't have any large enough jupyter notebook repo to make relevant tests